### ISSUE TITLE
Use the "local stageout manager" if the execution site==storage site

### ISSUE THREAD

**Issue by mmascher** [dmwm/CRABServer#5156]
In this case we use the "remote stageout manager" that simply does a gfal-copy instead of looking at what the site wants in SITECONF.

See https://hypernews.cern.ch/HyperNews/CMS/get/crabDevelopment/2497.html

**Comment by juztas** [comment]
Mh.. Looking at the code,  it might be more tricky. So cmscp initializes 2 stageout managers:
```
a) local (site specification)
    fallback (site specification) # This comes by default from site-local-config
b) remote (crab3 specification use srmv2-lcg and gfal-copy if it is available), which we call direct stageout....
```
if execution site == dest_site it will change the policy to do first remote, after that local. This check can be modified and if  execution site == dest_site, use local stageout policy not to temp, but final dest. We can`t use it always for remote, because if  execution site != dest_site and local, fallback failed,  local stageout can be configured do simple cp which we would use for direct, which will not work for sure. 

CRAB3 tries to be nice and skip ASO in this step, because this is the same source & dest site. I would suggest to have another option in site-local-config.xml, like direct-stage-out or remote-stage-out, but I suspect it would be always duplicate and the same as local stageout. If direct stageout is not defined in site-local-config, CRAB3 uses what it used before.

```
<direct-stage-out>
    <command value="vandy" />
    <catalog url="trivialcatalog_file://gpfs21/grid/grid-app/cmssoft/cms/SITECONF/local/PhEDEx/storage.xml?protocol=direct"/>
     <se-name value="se1.accre.vanderbilt.edu" />
     <phedex-node value="T2_US_Vanderbilt" />
</direct-stage-out>
```

**Comment by PerilousApricot** [comment]
Digging through, I understand the problem all the way. In retrospect, it seems like "local" and "remote" policies would be better named "asynchronous" and "synchronous".

The code wants to do something like this:

  | policy = local | Policy = remote
------------ | ------------- | ----------------------
source=dest | n/a | destLFN = permenant, manager = remote
source!=dest | destLFN = temp, manager = local | destLFN = permenant, manager = remote

And it hardcodes some assumptions about what "local" and "remote" are doing (which is what @juztas was talking about). What would be better is

  | policy = local | Policy = remote
------------ | ------------- | ----------------------
source=dest | destLFN = permanent, manager = local | destLFN = permanent, manager = remote
source!=dest | destLFN = temp, manager = local | destLFN = permanent, manager = remote

...but different parts of the code assume that if the policy was local, the file is in a temporary location. I'm putting together a patch that adds a "was the file moved to a permanent LFN" flag that propagates (eventually) through perform_stageout to the ASO injection step so it can select whether or not to inject to ASO depending on what the stageout did, and not just by policy.

FWIW - if there is gonna be hackery, I'd suggest it be in the tool (where it can be changed once for good) instead of asking every site to change their SITECONF.

**Comment by belforte** [comment]
@PerilousApricot thanks for having had the stomach to decode cmscp.py into that nice table :-(
I haven't looked at your commit, but I think (since long time) that we could use some refactoring there. 
Indeed clarity of intent comes first.
 Why use remote manager (no matter the policy) if source=dest ?
 How is it handled when local falls back to off-site storage ?
 Why do we write code before we have documented "the plan" ?

**Comment by PerilousApricot** [comment]
@belforte Yeah, I strongly resisted the urge to add a few more commits just to refactor everything. It was pretty crazy difficult to follow.

* Why use remote manager (no matter the policy) if source=dest ?
That behavior comes from #4892/#4894 and was modified by cbda09f6f7ed378ed48728548edf7750114910e9. Previously, the local policy (which is used 99% of the time) would use the local stageout to temp space, then trigger ASO injection to move the file from temp space to permanent space on the same SE. So, at that time, the table looked like this:

                | policy = local | Policy = remote
------------ | ----------------- | ----------------------
source=dest | destLFN = temp, manager = local | destLFN = permanent, manager = remote
source!=dest | destLFN = temp, manager = local | destLFN = permanent, manager = remote

The commits above changed the behavior if "policy = local" from "destLFN = temp, manager = local" to "destLFN = permanent, manager = remote". This patch backs it out halfway and changes the behavior to "destLFN = permanent, manager = local" 

* How is it handled when local falls back to off-site storage ?
perform_local_stageout() delegates the transfer and possible fallback to StageOutMgr, who passes back info about where the data was actually staged to CRAB. That dict is used to figure out what SE the file is "actually" at so ASO, the job report, etc.. can be updated. (though I just noticed I missed one change that I need to add)

* Why do we write code before we have documented "the plan" ?
Because it was Saturday night and I was waiting on some jobs to come back for my analysis :+1:. I assumed that since @mmascher took my question and turned it into an issue, it was ripe for a resolution. If someone wants to implement it differently, it won't hurt my feelings.  I'm not married to the code :)

**Comment by belforte** [comment]
I was not referring specifically to you when commenting about "write code before the plan". But this 
is a clear example where if the table of the desired behavior were explicit, a lot of time would 
have been saved :-)

**Comment by PerilousApricot** [comment]
I misunderstood :).

In any case, I need to figure out how to test this, which means hunting down how to set up a dummy crab WN environment on my machine.

**Comment by belforte** [comment]
I need to figure out exactly if this is still needed, then do it ASAP.

**Comment by PerilousApricot** [comment]
@belforte we still really want this. It's much better for our storage element and network to write from the worker nodes direct to the storage instead of going all the way through SRM (which is the whole reason we have a local storage plugin in the first place...)

**Comment by belforte** [comment]
Undrestood. Yes it is a shame that it is not used. OK. let's hope I can get to it next week
